File: LeetCode/Python/Medium/791-Custom-Sort-String/solution.py
from collections import Counter

class Solution:
    def customSortString(self, order: str, s: str) -> str:
        result=""

        hashmap=Counter(s)

        used=set()

        characters=[]

        for character in order:
            if character in hashmap.keys():
                characters.append(character * hashmap[character])

                del hashmap[character]

                used.add(character)

        for character in s:
            if character not in used:
                characters.append(character)

        result=''.join(characters)

        return result

class Solution:
    def customSortString(self, order: str, s: str) -> str:
        result=""

        hashmap=[0 for _ in range(0, 26)]

        for character in s:
            ascii=ord(character) - ord('a')

            hashmap[ascii]+=1

        used=[False for _ in range(0, 26)]

        characters=[]

        for character in order:
            ascii=ord(character) - ord('a')

            frequency=hashmap[ascii]

            if frequency > 0:
                characters.append(character * frequency)

                hashmap[ascii]=0

                used[ascii]=True

        for character in s:
            ascii=ord(character) - ord('a')
            
            if not used[ascii]:
                characters.append(character)

        result=''.join(characters)

        return result

File: LeetCode/Rust/Medium/791-Custom-Sort-String/solution.rs
use std::collections::{HashMap, HashSet};

impl Solution
{
    pub fn custom_sort_string(order: String, s: String) -> String
    {
        let mut result: String=String::from("");

        let mut hashmap: HashMap<char, i32>=HashMap::new();

        for character in s.chars()
        {
            hashmap.entry(character).and_modify(|frequency| *frequency+=1).or_insert(1);
        }

        let mut used: HashSet<char>=HashSet::new();

        for character in order.chars()
        {
            if hashmap.contains_key(&character)
            {
                let frequency: i32=*hashmap.get(&character).unwrap();

                for iteration in 0..frequency
                {
                    result.push(character);
                }

                used.insert(character);
            }
        }

        for character in s.chars()
        {
            if !used.contains(&character)
            {
                result.push(character);
            }
        }

        result
    }
}

impl Solution
{
    pub fn custom_sort_string(order: String, s: String) -> String
    {
        let mut result: String=String::from("");

        let mut hashmap: Vec<i32>=vec![0; 26];

        for character in s.chars()
        {
            let ascii: usize=character as usize - 'a' as usize;

            hashmap[ascii]+=1;
        }

        let mut used: Vec<bool>=vec![false; 26];

        for character in order.chars()
        {
            let ascii: usize=character as usize - 'a' as usize;

            let frequency: i32=hashmap[ascii];

            if frequency > 0
            {
                for iteration in 0..frequency
                {
                    result.push(character);
                }

                used[ascii]=true;
            }
        }

        for character in s.chars()
        {
            let ascii: usize=character as usize - 'a' as usize;

            if used[ascii]==false
            {
                result.push(character);
            }
        }

        result
    }
}